4주차 피드백 받은 부분 리팩터링과 빠진 기능 수정

silvertae opened the pull request:
## Description
- 미뤄왔던 리팩터링과 빠진 기능을 추가하였습니다. 

## Key changes
### 로그인 안한 유저가 동네설정을 할 수 있었던 버그 수정
https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/9209277e-31eb-4146-b22f-4aeb6bebfca6
### 판매상태에 따라 다른 드롭다운 표시
https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/02843705-ab7c-4400-86b5-e473da1cd115
### 상세페이지에서 삭제 시 원래 페이지로 리다이렉트 시키는 기능 추가
https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/1f6e668a-4865-4e0d-9cdf-dbcdf06cfae1
### 상세페이지 헤더 스크롤 위치에 따라 배경색 변경
https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/35713b17-68d5-445a-b04e-b1d4f96fe067

- authStore에 로그인 여부를 판단할 수 있는 authenticated 상태와 이를 set해주는 함수를 추가했습니다.
- `ProductItem`에 있던 드롭다운을 리뷰어 피드백을 반영해서 따로 `ProductItemDropdown` 컴포넌트로 분리했습니다.
- `ProductItem` 배지를 판매상태에 따라 다르게 보이도록 변경했습니다. 일단 판매중일때는 안 나오게 구현했습니다.
- ItemDetails에서 삭제 성공시 기존 페이지에 있던 페이지로 navigate 하도록 했습니다.
- Header에 새로운 props인 type과 isScrollTop과 이에 따른 스타일이 바뀌는 기능이 추가되었습니다.

## To reviewers
- 아직 다 확인하진 않았는데 제이도 authStore에 따로 isLogin 상태를 만드신 것 같더라구요. 저는 이 부분을 그냥 authticated라는 이름으로 로그인 로그아웃할 때 true, false만 바꿔주도록 단순하게 구현했는데 추후 합치면서 변경하면 될 것 같네요.

## Issue
- #94

silvertae committed:
feat #94: authStore에 authenticated 상태 및 set 함수 추가

silvertae committed:
Merge branch 'dev' into fix/#94-W4-Feedback

silvertae committed:
feat #94: 홈화면에서 로그인 된 유저가 아니면 동네설정이 불가능하도록 수정

silvertae committed:
feat #94: 상세 페이지에서 삭제 성공시 원래 있던 페이지로 리다이렉트 시키는 기능 추가

silvertae committed:
feat #94: ProductItem에 상품 삭제 Alert 추가

silvertae committed:
feat #94: ProductItem 판매상태에 따라 다른 Badge 렌더링

silvertae committed:
refactor #94: ProductItem에서 쓰이는 드롭다운 별도 컴포넌트로 분리 후 리팩터링

- ProductItemDropdown 컴포넌트로 분리
- 판매 상태에 따라 다른 메뉴를 보여주게 수정

silvertae committed:
chore #94: Home에서 불필요한 코드 제거

silvertae committed:
refactor #94: ...rest로 className을 받던 컴포넌트에 props 타입 명시

silvertae committed:
refactor #94: 상세페이지 StyledHeader 제거 및 Header 리팩터링

silvertae committed:
feat #94: 상세페이지 스크롤 위치에 따른 Header 색상변경 기능 추가

silvertae committed:
fix #94: Dropdown 버튼에 있었던 버그 수정

altmit commented on src/components/Alert.tsx:
```diff
@@ -20,6 +20,7 @@ export function Alert({ isOpen, onClose, onAction, children }: AlertProps) {
   const rootElement = document.getElementById('modal-root');
 
   const onAlertClose = (event: BaseSyntheticEvent) => {
+    event.stopPropagation();
```
어떤 이벤트 때문에 전파는 막는거죠?

silvertae replied:
`ProductItem` 컴포넌트에서 전체를 감싼 `Div`에 `onClick={() => showItemDetails(id)}`이 걸려있어서 이를 막기 위해 사용했습니다. 

Alert만 열었다 닫았을 뿐인데 자꾸 상품 상세페이지로 넘어가는 오류가 있어서 원인을 찾아봤는데 이게 문제였네요.

altmit commented on src/components/Badge.tsx:
```diff
@@ -15,25 +16,25 @@ export type BadgeProps = {
 };
 
 export function Badge({
```
Badge라는 컴포넌트를 styled-componenet로 상속받아 사용하기 위해 class name을 주는 것 보다 상속 없이 사용할 수 있는게 베스트라 생각하는데 그래서 변경한다면 다음 방법처럼 그냥 새로운 element로 한겹 더 쌓는게 깔끔할 것 같네요.

```tsx
// ItemDetails.tsx
<PageNav>
  <Badge
    fontColor="neutralTextWeak"
    badgeColor="neutralBackgroundBlur"
    text={`${activeSlidePage} / ${itemDetailsData.images.length}`}
    size="M"
    type="container"
  />
</PageNav>

  
//   ...
const PageNav = styled.div`
  position: absolute;
  bottom: 16px;
  right: 16px;
`;
```


className이라는게 styled-componenet로 `상속` 받을 때 만 필요한 값인데 그 외에는 전혀 필요 없기 때문에 만약 다르게 해결이 된다면(`지금 처럼 한겹 더 쌓는 방법 처럼`) 없는게 맞는것 같습니다.

만약 위 처럼 해결이 안돼서 필요하다면 상속해서 넣어줄 style을 차라리 옵션을 추가하는게 맞을 것 같네요.

silvertae replied:
어차피 위치를 잡기위해서 확장시켜 사용하던 부분이어서 이미 공통컴포넌트로 여러곳에서 사용되고 있는 Badge를 수정하기보다는 PageNav로 한 번 더 감싸는게 좋을 것 같네요.

altmit commented on src/components/Header.tsx:
```diff
@@ -1,51 +1,73 @@
+    `;
+  } else {
+    return css``;
+  }
```
그냥 else 없어도 될 것 같네요

altmit commented on src/components/dropdown/ProductItemDropdown.tsx:
```diff
@@ -0,0 +1,91 @@
+  };
+
+  return (
+    <Dropdown iconName="dots" align="right">
```
```ts
const statusActions = [
  { status: '판매중', action: changeStatusToOnSale, label: '판매중 상태로 전환' },
  { status: '예약중', action: changeStatusToReserved, label: '예약중 상태로 전환' },
  { status: '판매완료', action: changeStatusToSold, label: '판매완료 상태로 전환' }
];

// ...

<Dropdown iconName="dots" align="right">
  <MenuItem onAction={editItem}>게시글 수정</MenuItem>
  {statusActions.map(({ status: itemStatus, action, label }) => 
    status !== itemStatus && (
      <MenuItem key={itemStatus} onAction={action}>
        {label}
      </MenuItem>
    )
  )}
  <MenuItem color="systemWarning" onAction={openAlert}>
    삭제
  </MenuItem>
</Dropdown>
```
이렇게 작성한 다면 상태를 표현할 옵션들과 렌더링할 조건이 있는 로직이 분리되어서 아주 조금 더 괜찮지 않을까 싶긴하네요

altmit commented on src/page/ItemDetails.tsx:
```diff
@@ -45,9 +45,22 @@ export type ItemDetailsData = {
+  const [isScrollTop, setIsScrollTop] = useState(true);
+  const mainRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    window.addEventListener('wheel', onScroll);
+    window.addEventListener('touchmove', onScroll);
+    return () => {
+      window.removeEventListener('wheel', onScroll);
+      window.removeEventListener('touchmove', onScroll);
+    };
+  }, []);
```
접속한 디바이스를 확인해서 모바일/웹을 구분하여서 이벤트를 적용하면 어떨까요

silvertae replied:
괜찮은 방법인데 어차피 각 디바이스에서는 둘 중 하나의 이벤트만 일어나니까 이벤트 리스너가 걸려있더라도 실제 이벤트가 일어나지 않는다면 지금으로도 성능상 별 문제는 없을 것 같네요.

altmit replied:
둘 중 하나만 일어난다면 불필요한 이벤트를 메모리에 올려 놓을 필요 없지 않나요?

silvertae replied:
뭔가 접속한 디바이스를 확인하려면 복잡한 로직이 추가될 것 같아서 큰 문제없으면 그냥 두려고 했는데, `navigator.userAgent`에 정규식을 돌리는 간단한 방법이 있었네요.

추가) 근데 마냥 간단치는 않네요. 정규식 만으로 모든 모바일 디바이스와 웹을 정확히 구분하는데는 한계가 있다고 합니다. 그래서 아예 접속 디바이스를 구분하기 위한 라이브러리도 따로 있나봐요.

altmit replied:
정확한 구분은 안되는 것 인가요?
저는 정확하다고 알고 있었는데 아닌가 보군요.

altmit commented on src/page/ItemDetails.tsx:
```diff
@@ -68,6 +81,14 @@ export function ItemDetails() {
+      setIsScrollTop(true);
+    } else {
+      setIsScrollTop(false);
+    }
```
```suggestion
   const scrollTop = mainRef.current?.scrollTop === 0
   setIsScrollTop(scrollTop );
```

altmit commented on src/components/Header.tsx:
```diff
@@ -1,51 +1,73 @@
 
 export function Header({
+  type = 'default',
+  isScrollTop,
```
`isScrollTop`은 Header가 알 필요 없는 관심사 인 것 같습니다.
type을 외부에서 지정해야지 외부 props에 따라서 스스로 지정하는 건 이상한 것 같습니다.

> 외부의 값에 의해서 결정되는 값을 전달해서 그것에 따라 값을 type을 수정 시킨다


> 외부에서 type을 지정한다.

header가 수행해야 할 논리적 연산이 하나 사라지는 것 같네요.

altmit replied:
그렇다면 type도 수정이 필요하 겠네요.

default와 dynamic이 아닌

backdropFilter가 적용된 상태와 그렇지 않은 상태로

```ts
type = 'default' | 'backdrop'
// default : display block, backdrop 미적용, background 적용
// backdrop : display absolute, backdrop 적용, background 미적용
```

```ts
<Header type={isScrollTop? 'backdrop' : 'default'} />
```

이렇게 사용 한다면 Header가 직접 연산할 필요가 없어지며 스크롤 여부에 따라서 투명한 배경이 되었다, 자리 차지하는 상태가 될 것 같네요

silvertae replied:
제이 말씀처럼 `isScrollTop`을 Header의 외부에서 연산해서 타입을 지정해주도록 리팩터링 한다면 type 3개가 필요할 것 같네요.

```tsx
// default는 상세페이지 제외 모든 페이지

// default: backdrop 적용, background-color: 있음
// background: backdrop 미적용, background-color: transparent
// transparent: backdrop 미적용, background-color: 있음

...
// 상세페이지
<Header type={ isScrollTop ? 'transparent' : 'background' } />
```
그리고 display는 absolute를 계속 유지해도 될 것 같습니다.

altmit replied:
type이 3개까지 필요한가요?

투명 배경과 그렇지 않은 경우만 있어도 될 것 같은 기분인데🤔

silvertae replied:
뭐야 기획서에 분명 상단 헤더가 첫 진입시 투명하다가 스크롤을 하면 backdrop-filter 없는 배경이 생기는거 아니였나요?
기획서 다시보니까 왜 불투명이라고 써있지... 전 그동안 뭘 본거죠? 😂 ㅋㅋ

altmit replied:
음 저도 계속 투명이라 언급해서 그런걸지도..?ㅋㅋㅋㅋ
암튼 색상이 있는 경우와 backdrop-filter가 적용된 경우 두가지가 맞을까요?

altmit commented on src/page/ItemDetails.tsx:
```diff
@@ -45,9 +45,22 @@ export type ItemDetailsData = {
   const { itemId } = useParams();
   const navigate = useNavigate();
+  const location = useLocation();
+  const from = location?.state?.redirectedFrom.pathname || '/';
```
아마 이 부분 리뷰어가 명칭이 referer라고 말씀해 주셨던 것 같네요

silvertae replied:
저도 찾아봤는데 referrer 는 아예 다른 페이지에서 하이퍼링크를 통해 넘어왔을 때 의미가 있는 값인 것 같더라구요.
지금 저희 앱에서 `document.referrer`를 콘솔에 찍어보면 어떤 페이지든 그냥 `http://localhost:5173/`만 찍힙니다.

지금처럼 react-router를 이용해서 navigate로 이동하는 경우에는 지금 방식이 알맞은 것 같네요.

altmit commented on src/api/fetchers/authFetcher.ts:
```diff
@@ -19,13 +20,23 @@ export const singup = async (formData: FormData) => {
     const res = await fetcher.post(API_ENDPOINT.LOGIN, loginInfo);
 
     if (res.status === 200) {
+      setAuthentication(true);
```
isLogin과 같이 로그인 여부를 확인하는 값 같은데

서버에서의 요청을 믿어야 하지만 status 200만으론 로그인 되었다고 확신까진 못할 거 같네요.

저희 입장에서는 값이 하나라도 비어 있거나 하면 로그인 처리를 안 해야 할 것 같네요.
일종의 더블 체크라 생각합니다.

silvertae replied:
그렇다면 서버에서는 성공했지만 프론트에서 누락된 데이터가 있으면 서버와 사용자에게 이에 대해 알려주는 에러 핸들링이 추가적으로 필요할 것 같네요.

altmit commented on src/page/home/Home.tsx:
```diff
@@ -93,28 +106,39 @@ export function Home() {
-              })}
-              <MenuItem onAction={openModal}>내 동네 설정하기</MenuItem>
-            </Dropdown>
+            {!authenticated ? (
+              <Button
+                styledType="text"
+                color="neutralText"
+                onClick={onLocationClick}
+              >
+                역삼1동
+                <Icon name="chevronDown" color="neutralTextStrong" />
+              </Button>
+            ) : (
+              <Dropdown
+                btnText={
+                  extractKeyName(itemData?.pages[0]?.userLocation) || '역삼1동'
+                }
+                iconName="chevronDown"
+                align="left"
+              >
+                {userLocationData?.locations?.map(location => {
+                  return (
+                    <MenuItem
+                      key={location.id}
+                      isSelected={location.isSelected}
+                      onAction={() => {
+                        selectLocation(location.id, location.isSelected);
+                      }}
+                    >
+                      {location.name}
+                    </MenuItem>
+                  );
+                })}
+                <MenuItem onAction={openModal}>내 동네 설정하기</MenuItem>
+              </Dropdown>
+            )}
```
사이즈가 너무 비대해 졌네요.

컴포넌트로 분리하는 방법도 있고 아니면 dropdown을 리팩토링해서 지금처럼 조건부 렌더링이 아닌 로그인 여부에 따라서 클릭 안되고 toast를 띄우는 컴포넌트로 간소화도 가능할 것 같네요

silvertae replied:
원래 Dropdown 내부에서 로그인 여부를 판단해서 토스트를 띄우도록 했다가 그렇게 되면 드롭다운 버튼이 어떤 컴포넌트에서 눌렸는지에 따라 다른 토스트 문구를 출력해줘야해서 지금처럼 바꾸었습니다 😅

아예 ProductItem에서처럼 Dropdown 자체를 분리하는 것도 좋을 것 같네요.

silvertae committed:
refactor #94: 리뷰어 피드백 반영

- Badge 컴포넌트 props에서 className 제거
- 불필요한 else 문 제거
- ProductItemDropdown에서 MenuItem 렌더링하는 방식 변경
- 상세페이지 디바이스에 따라 다른 이벤트 리스너 등록 및 제거
- Header 컴포넌트 props 수정

silvertae committed:
refactor #94: Home에 있던 드롭다운 별도 컴포넌트로 분리

- UserLocationDropdown 이라는 컴포넌트로 분리했습니다.
- extractKeyName 함수 유틸화

silvertae committed:
Merge branch 'dev' into fix/#94-W4-Feedback
